File: services/rust-actix/src/sample/second.rs
use actix_web::{HttpRequest, HttpResponse, Responder, web};
use chrono::{DateTime, Utc};
use itertools::Itertools;
use serde::Serialize;

use crate::sample::backend::{BackendClient, Moon};
use crate::sample::error::ErrorResponse;
use crate::shared::auth::auth;

#[derive(Serialize)]
struct Response {
    relevant: bool,
    omit: bool,
    description: String,
    weight: u64,
    items: Vec<Item>,
}

#[derive(Serialize)]
struct Item {
    name: String,
    details: Option<String>,
    timestamp: DateTime<Utc>,
    count: u64,
}

pub async fn second(
    path: web::Path<String>,
    backend_client: web::Data<BackendClient>,
    req: HttpRequest,
) -> impl Responder {
    let optional_current_user = auth(req);
    if optional_current_user.is_none() {
        return HttpResponse::Forbidden().json(ErrorResponse { message: String::from("Bad authorization") });
    }
    let current_user = optional_current_user.unwrap();
    let user_token = current_user.raw_token();
    let id = path.into_inner();

    let fetch_medium_result = backend_client.fetch_medium(
        &user_token,
        &format!("{}++sec-{}", current_user.id(), id),
    ).await;
    if let Err(response) = fetch_medium_result {
        return response;
    }
    let planet = fetch_medium_result.unwrap();
    let name1 = get_name(&planet.moons, 0);
    let name2 = get_name(&planet.moons, 1);
    let name3 = get_name(&planet.moons, 2);

    let fetch_small_result1 = backend_client.fetch_small(
        &user_token,
        &format!("foo_{}", name1),
    ).await;
    if let Err(response) = fetch_small_result1 {
        return response;
    }
    let opera1 = fetch_small_result1.unwrap();

    let fetch_small_result2 = backend_client.fetch_small(
        &user_token,
        &format!("bar_{}", name2),
    ).await;
    if let Err(response) = fetch_small_result2 {
        return response;
    }
    let opera2 = fetch_small_result2.unwrap();

    let fetch_small_result3 = backend_client.fetch_small(
        &user_token,
        &format!("quz_{}", name3),
    ).await;
    if let Err(response) = fetch_small_result3 {
        return response;
    }
    let opera3 = fetch_small_result3.unwrap();

    HttpResponse::Ok().json(Response {
        relevant: planet.missions.unwrap_or(vec![]).iter().any(|m| m.contains("f")),
        omit: !planet.gas,
        description: planet.moons.iter().map(|moon| &moon.name).join("--"),
        weight: planet.diameter + planet.orbit,
        items: vec![
            Item{
                name: name1.to_string(),
                details: opera1.style,
                timestamp: opera1.composed_at,
                count: opera1.number_of_acts,
            },
            Item{
                name: name2.to_string(),
                details: opera2.style,
                timestamp: opera2.composed_at,
                count: opera2.number_of_acts,
            },
            Item{
                name: name3.to_string(),
                details: opera3.style,
                timestamp: opera3.composed_at,
                count: opera3.number_of_acts,
            },
        ],
    })
}

fn get_name(moons: &Vec<Moon>, index: usize) -> &str {
    if index >= moons.len() {
        return "(none)";
    }
    return &moons[index].name;
}

File: services/go-gin/internal/sample/second.go
package sample

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"sample-micro-service/pkg/service/auth"
	"strings"
	"time"
)

type SecondResponse struct {
	Relevant    bool         `json:"relevant"`
	Omit        bool         `json:"omit"`
	Description string       `json:"description"`
	Weight      int          `json:"weight"`
	Items       []SecondItem `json:"items"`
}

type SecondItem struct {
	Name      string    `json:"name"`
	Details   *string   `json:"details"`
	Timestamp time.Time `json:"timestamp"`
	Count     int       `json:"count"`
}

func Second(c *gin.Context) {
	userId := c.GetString(auth.UserId)
	userToken := c.GetString(auth.RawToken)
	id := c.Param("id")

	planet, success := fetchMedium(c, userToken, userId+"++sec-"+id)
	if !success {
		return
	}
	name1 := getName(planet.Moons, 0)
	name2 := getName(planet.Moons, 1)
	name3 := getName(planet.Moons, 2)

	opera1, success := fetchSmall(c, userToken, "foo_"+name1)
	if !success {
		return
	}
	opera2, success := fetchSmall(c, userToken, "bar_"+name2)
	if !success {
		return
	}
	opera3, success := fetchSmall(c, userToken, "quz_"+name3)
	if !success {
		return
	}

	relevant := false
	for _, mission := range planet.Missions {
		if strings.Contains(mission, "f") {
			relevant = true
			break
		}
	}

	description := ""
	for _, moon := range planet.Moons {
		if len(description) > 0 {
			description += "--"
		}
		description += moon.Name
	}

	c.JSON(http.StatusOK, SecondResponse{
		Relevant:    relevant,
		Omit:        !planet.Gas,
		Description: description,
		Weight:      planet.Diameter + planet.Orbit,
		Items: []SecondItem{
			{
				Name:      name1,
				Details:   opera1.Style,
				Timestamp: opera1.ComposedAt,
				Count:     opera1.NumberOfActs,
			},
			{
				Name:      name2,
				Details:   opera2.Style,
				Timestamp: opera2.ComposedAt,
				Count:     opera2.NumberOfActs,
			},
			{
				Name:      name3,
				Details:   opera3.Style,
				Timestamp: opera3.ComposedAt,
				Count:     opera3.NumberOfActs,
			},
		},
	})
}

func getName(moons []Moon, index int) string {
	if index >= len(moons) {
		return "(none)"
	}
	return moons[index].Name
}
